core/services/ocr2/plugins/threshold/decryption_queue.go

```diff
@@ -0,0 +1,226 @@
+package decryption_queue
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/job"
+)
+
+type CiphertextId = []byte
+
+type ThresholdDecryptor interface {
+	Decrypt(ctx context.Context, ciphertextId CiphertextId, ciphertext []byte) ([]byte, error)
+}
+
+// This interface will be replaced by the Threshold Decryption Plugin when it is ready
+type DecryptionQueuingService interface {
+	GetRequests(requestCountLimit uint32, totalBytesLimit uint32) []DecryptionRequest
+	GetCiphertext(ciphertextId CiphertextId) ([]byte, error)
+	ResultReady(ciphertextId CiphertextId, plaintext []byte)
+}
+
+type DecryptionRequest struct {
+	ciphertextId CiphertextId
+	ciphertext   []byte
+}
+
+type pendingRequest struct {
+	chPlaintext chan []byte
+	ciphertext  []byte
+}
+
+type completedRequest struct {
+	plaintext []byte
+	timer     *time.Timer
+}
+
+type decryptionQueue struct {
+	maxQueueLength                uint32
+	maxCiphertextBytes            uint32
+	completedRequestsCacheTimeout time.Duration
+	pendingRequestQueue           []CiphertextId
+	pendingRequests               map[string]pendingRequest
+	completedRequests             map[string]completedRequest
+	mu                            sync.RWMutex
+	lggr                          logger.Logger
+}
+
+var (
+	_ ThresholdDecryptor       = &decryptionQueue{}
+	_ DecryptionQueuingService = &decryptionQueue{}
+	_ job.ServiceCtx           = &decryptionQueue{}
+)
+
+func NewThresholdDecryptor(maxQueueLength uint32, maxCiphertextBytes uint32, completedRequestsCacheTimeoutMs uint64, lggr logger.Logger) ThresholdDecryptor {
+	dq := decryptionQueue{
+		maxQueueLength,
+		maxCiphertextBytes,
+		time.Duration(completedRequestsCacheTimeoutMs) * time.Millisecond,
+		[]CiphertextId{},
+		make(map[string]pendingRequest),
+		make(map[string]completedRequest),
+		sync.RWMutex{},
+		lggr,
+	}
+	return &dq
+}
+
+func (dq *decryptionQueue) Decrypt(ctx context.Context, ciphertextId CiphertextId, ciphertext []byte) ([]byte, error) {
+	if len(ciphertext) > int(dq.maxCiphertextBytes) {
+		return nil, errors.New("ciphertext too large")
+	}
+
+	chPlaintext, err := dq.getResult(ciphertextId, ciphertext)
+	if err != nil {
+		return nil, err
+	}
+
+	select {
+	case pt, ok := <-chPlaintext:
+		if ok {
+			return pt, nil
+		}
+		return nil, fmt.Errorf("pending decryption request for ciphertextId %s was closed without a response", string(ciphertextId))
+	case <-ctx.Done():
+		dq.mu.Lock()
+		defer dq.mu.Unlock()
+		delete(dq.pendingRequests, string(ciphertextId))
+		return nil, errors.New("context provided by caller was cancelled")
+	}
+}
+
+func (dq *decryptionQueue) getResult(ciphertextId CiphertextId, ciphertext []byte) (chan []byte, error) {
```

Comment:
Who owns the channel? Can we return receive-only from here?
```suggestion
func (dq *decryptionQueue) getResult(ciphertextId CiphertextId, ciphertext []byte) (<-chan []byte, error) {
```